components/script/dom/rtcpeerconnection.rs

```diff
@@ -256,6 +292,65 @@ impl RTCPeerConnection {
         event.upcast::<Event>().fire(self.upcast());
     }
 
+    fn on_data_channel_event(&self, channel_id: DataChannelId, event: DataChannelEvent, global: &GlobalScope) {
+        if self.closed.get() {
+            return;
+        }
+
+        match event {
+            DataChannelEvent::NewChannel => {
+                let channel = RTCDataChannel::new(
+                    &self.global(),
+                    &self,
+                    USVString::from("".to_owned()),
+                    &RTCDataChannelInit::empty(),
+                    Some(channel_id),
+                );
+
+                let event = RTCDataChannelEvent::new(
+                    &self.global(),
+                    atom!("datachannel"),
+                    false,
+                    false,
+                    &channel,
+                );
+                event.upcast::<Event>().fire(self.upcast());
+            },
+            _ => {
+                let channel = if let Some(channel) = self.data_channels.borrow().get(&channel_id) {
+                    DomRoot::from_ref(&**channel)
+                } else {
+                    debug_assert!(false, "Got an event for an unregistered data channel");
```

Comment:
I hit this assert when testing when I pressed the "stop" button immediately after starting it, since a bunch of other data channel events were received afterwards. A `warn!` might be more appropriate.

Reply:
Actually I seem to hit it when I only click the start button and nothing else. I added some debug output and saw this:
```
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "StateChange" for 0
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 0
Send channel state is: open
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "NewChannel" for 0
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] registering channel 0
Receive Channel Callback
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 0
Receive channel state is: connecting
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "NewChannel" for 1
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] registering channel 1
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 1
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "NewChannel" for 2
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] registering channel 2
Receive Channel Callback
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 2
Receive channel state is: connecting
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 3
[2020-06-23T19:54:28Z WARN  script::dom::rtcpeerconnection] Got an event for an unregistered data channel (3)
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "NewChannel" for 3
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] registering channel 3
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 3
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 4
[2020-06-23T19:54:28Z WARN  script::dom::rtcpeerconnection] Got an event for an unregistered data channel (4)
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "NewChannel" for 4
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] registering channel 4
Receive Channel Callback
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 4
Receive channel state is: connecting
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 5
[2020-06-23T19:54:28Z WARN  script::dom::rtcpeerconnection] Got an event for an unregistered data channel (5)
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "NewChannel" for 5
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] registering channel 5
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 5
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "Open" for 6
[2020-06-23T19:54:28Z WARN  script::dom::rtcpeerconnection] Got an event for an unregistered data channel (6)
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] processing "NewChannel" for 6
[2020-06-23T19:54:28Z DEBUG script::dom::rtcpeerconnection] registering channel 6
```

Reply:
I have a suspicion that this comes from https://github.com/servo/media/blob/73b2e5485c1f6384adf068901cb4ac771c945540/backends/gstreamer/webrtc.rs#L564, which looks like it could cause us to get an Open event for a channel before we receive the NewChannel event for it.